File: src/main/java/pl/training/workshop/m7/s01_breakdependencies/step3/ShowtimeReminderJob.java
package pl.training.workshop.m7.s01_breakdependencies.step3;

import java.time.Clock;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Objects;
import java.util.function.Supplier;

import pl.training.workshop.m7.s01_breakdependencies.PaidBooking;

/**
 * Krok 3: Subclass and Override Method - wywołanie statycznego mailera przeniesione
 * do metody chronionej sendReminder(). Test nadpisuje ją w podklasie (seam) i po raz
 * pierwszy widzi wysłane przypomnienia. Klasa przestała być final - to cena tego seamu.
 */
public class ShowtimeReminderJob {
    private final Supplier<BookingStore> stores;
    private final Clock clock;

    public ShowtimeReminderJob() {
        this(LegacyDatabase::new, Clock.systemDefaultZone());
    }

    public ShowtimeReminderJob(Supplier<BookingStore> stores, Clock clock) {
        this.stores = Objects.requireNonNull(stores, "stores");
        this.clock = Objects.requireNonNull(clock, "clock");
    }

    public int run() {
        BookingStore database = stores.get();
        LocalDateTime now = LocalDateTime.now(clock);
        int sent = 0;
        for (PaidBooking booking : database.paidBookings()) {
            long minutes = Duration.between(now, booking.start()).toMinutes();
            if (!booking.reminded() && minutes > 0 && minutes <= 120) {
                sendReminder(booking.email(),
                        "Przypomnienie: " + booking.title(),
                        "Seans zaczyna sie o " + booking.start().toLocalTime());
                database.markReminded(booking.id());
                sent++;
            }
        }
        return sent;
    }

    protected void sendReminder(String to, String subject, String body) {
        ReminderMailer.send(to, subject, body);
    }
}

File: typescript/src/workshop/m7/s01_breakdependencies/step3/ShowtimeReminderJob.ts
import { requireNonNull } from '../../../../shared/requireNonNull.js';
import { type Clock, Duration, LocalDateTime, systemClock } from '../../../shared/time.js';
import type { BookingStore } from './BookingStore.js';
import { LegacyDatabase } from './LegacyDatabase.js';
import { ReminderMailer } from './ReminderMailer.js';

/**
 * Krok 3: Subclass and Override Method - wywołanie statycznego mailera przeniesione
 * do metody chronionej sendReminder(). Test nadpisuje ją w podklasie (seam) i po raz
 * pierwszy widzi wysłane przypomnienia. Klasa ma teraz punkt rozszerzenia przez
 * dziedziczenie (protected) - to cena tego seamu.
 */
export class ShowtimeReminderJob {
  private readonly stores: () => BookingStore;
  private readonly clock: Clock;

  constructor(stores: () => BookingStore = () => new LegacyDatabase(), clock: Clock = systemClock) {
    this.stores = requireNonNull(stores, 'stores');
    this.clock = requireNonNull(clock, 'clock');
  }

  run(): number {
    const database = this.stores();
    const now = LocalDateTime.now(this.clock);
    let sent = 0;
    for (const booking of database.paidBookings()) {
      const minutes = Duration.between(now, booking.start).toMinutes();
      if (!booking.reminded && minutes > 0 && minutes <= 120) {
        this.sendReminder(booking.email,
          `Przypomnienie: ${booking.title}`,
          `Seans zaczyna sie o ${booking.start.toLocalTime().toString()}`);
        database.markReminded(booking.id);
        sent++;
      }
    }
    return sent;
  }

  protected sendReminder(to: string, subject: string, body: string): void {
    ReminderMailer.send(to, subject, body);
  }
}
